rclcpp/src/rclcpp/executors/static_single_threaded_executor.cpp

```diff
@@ -18,321 +18,130 @@ using rclcpp::executor::ExecutableList;
 
 StaticSingleThreadedExecutor::StaticSingleThreadedExecutor(
   const rclcpp::executor::ExecutorArgs & args)
-: executor::Executor(args) {}
+: executor::Executor(args)
+{
+  entities_collector_ = std::make_shared<StaticExecutorEntitiesCollector>();
+}
 
 StaticSingleThreadedExecutor::~StaticSingleThreadedExecutor() {}
 
-
 void
 StaticSingleThreadedExecutor::spin()
 {
   if (spinning.exchange(true)) {
     throw std::runtime_error("spin() called while already spinning");
   }
   RCLCPP_SCOPE_EXIT(this->spinning.store(false); );
-  rclcpp::executor::ExecutableList executable_list;
-  run_collect_entities();
-  get_executable_list(executable_list);
+
+  // Set memory_strategy_ and exec_list_ based on weak_nodes_
+  // Prepare wait_set_ based on memory_strategy_
+  entities_collector_->init(&wait_set_, memory_strategy_);
+
   while (rclcpp::ok(this->context_) && spinning.load()) {
-    execute_ready_executables(executable_list);
+    execute_ready_executables();
```

Comment:
can you move `entities_collector_->refresh_wait_set(timeout);` out of `execute_ready_executables` and put it here?